### ISSUE TITLE
add depending option relationship

### ISSUE THREAD

**Issue by hsn10** [scallop/scallop#64]
last thing is option relationship are depending options

for example for creating filesystem you can specify optional filesystem label. You will have options like this

-c .... create filesystem operation
-u .... update filesystem operation
-d ... destroy filesystem operation
-L <name> ... filesystem name

you can use -c without specifying -L, but you can use -L only if -c or -u is also used

proposed syntax:

depending(parent, child)

**Comment by Rogach** [comment]
It's already possible:

``` scala
object Conf extends ScallopConf(args) {
  val create = opt[Boolean]()
  val name = opt[String]()
  validateOpt (create,name) {
    case (Some(false), Some(_)) => Left("Name option is unused")
    case _ => Right(Unit)
  }
}
```

**Comment by hsn10** [comment]
this is not very user friendly and does not helps with #63 . If this get implemented you can use this info for building better help message.

**Comment by Rogach** [comment]
Yes, it's not extremely user friendly, but it solves the problem in general way. Your proposal, on the other hand, will not work in the described situation - for example, if you define option relationship as in the following snippet:

``` scala
depending(create, name)
depending(update, name)
```

and call `create`, with name, it will fail - since name also depends on `update`.

**Comment by Rogach** [comment]
My point is, there is an infinite amount of possible relationships between options, and trying to catalogue and implement all of them will only bloat the library and confuse the users.

That said, if you would be able to implement something like this properly, I'll happily accept a pull request :)

**Comment by hsn10** [comment]
Dependency management is common task. Two depend modes should be sufficient to construct graph. As i see it you can depend on one item from list, or on all items on list. What case i am missing?

**Comment by Rogach** [comment]
Hm. I like the idea about the graph.

But two modes you suggest wouldn't be enough. First of all, depending on all items in the list is a special case of depending on one item from list - you just apply dependency several times (it still may be helpful to special-case that mode, though - to get better error messages):

``` scala
dependsOnAll(A, List(B,C,D))
// equals to:
dependsOnAny(A, List(B))
dependsOnAny(A, List(C))
dependsOnAny(A, List(D))
```

Currently, we have two helpers for managing dependencies - `codependent` and `mutuallyExclusive`. First one can be implemented in terms of `dependsOnAny`, second one is a bit trickier. To implement `mutuallyExclusive`, we'll also need some sort of anti-dependency - providing one option forbids usage of other option.

So I'll propose the following structure:
1. `dependsOnAny` - `A` can be provided iff at least one option in list is present
2. `dependsOnAll` - `A` can be provided iff all of options in list are present
3. `conflicts` - `A` can be provided iff none of options in list are present
4. `codependent` - either all or none options in list must be provided
5. `mutuallyExclusive` - either one or none options from list can be present

Cases (2), (4) and (5) are special cases, but we`ll need them to provide sane error messages to the user.

What do you think? Did I miss something?

P.S. I don't like names `dependsOnAll` and `dependsOnAny`, though. Maybe there's some better names for them?

**Comment by stevedlawrence** [comment]
I could use something similar to mutuallyExclusive, but with the restriction that one must be available, so something like:

`requireOne` - one and only one of the options from the list must be present

**Comment by Rogach** [comment]
@stevedlawrence - so you program can't be called with empty argument list?

**Comment by stevedlawrence** [comment]
Right, my simplified usage essentially looks like:

`foo (-a <string>... | -b <string>)`

So you must provide either a bunch of a's, or a single b. I thought about doing this separation with subcommands, but I already use subcommands so things would get a bit more messy than they need to be.

If you'd like to take a look, the project is [Daffodil](https://opensource.ncsa.illinois.edu/confluence/display/DFDL/Daffodil%3A+Open+Source+DFDL) and we have a full description of our [CLI](https://opensource.ncsa.illinois.edu/confluence/display/DFDL/Command+Line+Interface)

**Comment by Rogach** [comment]
@stevedlawrence - So, while we're fleshing the implementation for "new" option dependencies, here's the temporary workaround for you:

``` scala
valdiateOpt(optA, optB) {
  case (Some(a), Some(b)) => Left("Error")
  case (None, None) => Left("Error")
  case _ => Right(Unit)
}
```

**Comment by stevedlawrence** [comment]
Yeah, that's what I use right now (only a bit more complicated because of all the other options I have). And it works great. I was just saying if we wanted to cover all the cases for dependency management, I'd also needed something like requireOne.

**Comment by Rogach** [comment]
Sure, I'll add it to the list.

**Comment by hsn10** [comment]
its possible to use scala generics to have syntax like this?

``` scala
depends[Some](option, List) - some item from list is enough
depends[None[(option, list) - can not exists together
depends[All]...
```

**Comment by Rogach** [comment]
No, not possible - at least because type `Some` requires type parameter. What would be the benefit of such syntax?

**Comment by hsn10** [comment]
Its easier to read. Instead of _Some_ what about _Any_ ?

**Comment by Rogach** [comment]
It's easier to read for you maybe, but if you look deeper, you open a nice can of worms. At least, consider the (possible) type signature:

``` scala
def depends[A](opt: ScallopOption[_], list: List[ScallopOption[_]])(implicit val p: DependsParam[A])
```

It's quite confusing. And it's definitely _not_ easier to read. There was a lot of confusion in stdlib over scala.collection methods - for example, List.map - but there, complexity was justified. Here it is not justified, in my opinion.

After all, `dependsSome` is even more readable then `depends[Some]`. What's wrong with it?

P.S. `depends[None]` is also funny - if you read it literally, it means 'option doesn't depend on anything in the list'. Doesn't make sense - thus, less _readable_.

**Comment by hsn10** [comment]
I do not think about scallop in method signatures. For me, its more DSL for parsing options

**Comment by Rogach** [comment]
@hsn10 - Sure, but the same can be said about scala.collection - it's just a dsl for working with collections. And if people will use the dsl, they'll need to understand what's going on, ultimately.

**Comment by Rogach** [comment]
So, I implemented new option dependencies (dependsOnAny, dependsOnAll, conflicts, requireOne), and published a snapshot (for scala 2.10.x):

``` scala
resolvers += "sonatype snapshots" at "https://oss.sonatype.org/content/repositories/snapshots"
libraryDependencies += "org.rogach" %% "scallop" % "0.9.2~5~SNAPSHOT"
```

@hsn10, @stevedlawrence  - would be nice if you could try it out! :)
